src/components/virtualRepeater/virtualRepeater.js

```diff
@@ -0,0 +1,452 @@
+};
+
+
+VirtualRepeatContainerController.prototype.handleScroll_ = function(horizontal) {
+  var NUM_EXTRA = 3;
+
+  var transform, offset;
+  if (horizontal) {
+    offset = this.scroller.scrollLeft;
+    transform = 'translateX(';
+  } else {
+    offset = this.scroller.scrollTop;
+    transform = 'translateY(';
+  }
+
+  if (offset === this.scrollOffset) return;
+  this.scrollOffset = offset;
+
+  var itemSize = this.repeater.getItemSize();
+  transform += Math.max(0,
+      Math.floor(this.scrollOffset / itemSize) - NUM_EXTRA) * itemSize + 'px)';
+
+  this.offsetter.style.webkitTransform = transform;
+  this.offsetter.style.transform = transform;
+
+  this.repeater.containerUpdated();
+};
+
+
+/**
+ * @ngdoc directive
+ * @name mdVirtualRepeat
+ * @module material.components.virtualRepeat
+ * @restrict A
+ * @description
+ * `md-virtual-repeat` specifies an element to repeat using virtual scrolling.
+ *
+ * Virtual repeat is a limited substitute for ng-repeat that allows renders only
+ * enough dom nodes to fill the container and recycling them as the user scrolls.
+ *
+ * @usage
+ * <hljs lang="html">
+ * <md-virtual-repeat-container md-horizonal>
+ *   <div md-virtual-repeat="i in items">Hello {{i}}!</div>
+ * </md-virtual-repeat-containr>
+ *
+ * <md-virtual-repeat-container md-horizonal>
+ *   <div md-virtual-repeat="i in items">Hello {{i}}!</div>
+ * </md-virtual-repeat-containr>
+ * </hljs>
+ *
+ * @param {number=} md-size The height or width of the repeated elements (which
+ *     must be identical for each element). TODO: If absent, the directive will attempt to
+ *     measure the computed style of the element at startup.
+ */
+function VirtualRepeatDirective($parse) {
+  return {
+    controller: VirtualRepeatController,
+    priority: 1000,
+    require: ['mdVirtualRepeat', '^mdVirtualRepeatContainer'],
+    restrict: 'A',
+    terminal: true,
+    transclude: 'element',
+    compile: function VirtualRepeatCompile($element, $attrs) {
+      var expression = $attrs.mdVirtualRepeat;
+      var match = expression.match(/^\s*([\s\S]+?)\s+in\s+([\s\S]+?)\s*$/);
+      var lhs = match[1];
+      var rhs = $parse(match[2]);
+      var extraName = $attrs.mdExtraName && $parse($attrs.mdExtraName);
+
+      return function VirtualRepeatLink($scope, $element, $attrs, ctrl, $transclude) {
+        ctrl[0].link_(ctrl[1], $transclude, lhs, rhs, extraName);
+      };
+    }
+  };
+}
+
+
+function VirtualRepeatController($scope, $element, $attrs, $browser, $document) {
+  this.$scope = $scope;
+  this.$element = $element;
+  this.$attrs = $attrs;
+  this.$browser = $browser;
+  this.$document = $document;
+
+  this.browserCheckUrlChange = $browser.$$checkUrlChange;
+  this.newStartIndex = 0;
+  this.newEndIndex = 0;
+  this.startIndex = 0;
+  this.endIndex = 0;
+  // TODO: measure width/height of first element from dom if not provided.
+  // getComputedStyle?
+  this.itemSize = $scope.$eval($attrs.mdSize);
+  this.blocks = {};
```

Comment:
You use `blocks` a lot throughout the code, but it's not really explained what a `block` is in this context.